nn/transformer.py

```diff
@@ -0,0 +1,301 @@
+    x = self.linear_out(x)
+    x = nn.dropout(x, dropout=self.dropout)
+    return x
+
+
+class TransformerEncoder(nn.Module):
+  """
+  Defines the full Encoder of the standard transformer
+  """
+  def __init__(self, encoder_layer: Union[TransformerEncoderLayer, Any], num_layers: int,
+               norm=nn.layer_norm, norm_eps: float = 1e-5):
+    """
+    :param encoder_layer: Encoder layer to be stacked num_layers times
+    :param num_layers: Number of layers
+    :param norm: normalization function
+    :param norm_eps: Epsilon value for layer normalization
+    """
+    super().__init__()
+    import copy
+    self.layers = nn.ModuleList([copy.deepcopy(encoder_layer) for _ in range(num_layers)])
+
+    self.num_layers = num_layers
+    self.norm = norm
+    self.norm_eps = norm_eps
+
+  def forward(self, x: nn.LayerRef) -> nn.LayerRef:
+    """
+    Applies every encoder layer initialized in self.layers.
+    """
+    output = x
+
+    for mod in self.layers:
+      output = mod(output)
+
+    if self.norm is not None:
+      output = self.norm(output, epsilon=self.norm_eps)
+
+    return output
+
+
+class TransformerDecoderLayerStep(nn.Module):
+  """
+  Defines one layer of a standard transformer decoder
+  """
+  def __init__(self, dim_model: int, num_heads: int, dim_ff: int = 2048, dropout: float = 0.1, att_dropout: float = 0.1,
+               activation: Callable[[nn.LayerRef], nn.LayerRef] = nn.relu, norm_eps: float = 1e-5,
+               norm_first: bool = False, norm=nn.layer_norm):
+    """
+    :param dim_model: hidden dim, PyTorch name: d_model
+    :param num_heads: number heads, PyTorch name: nhead
+    :param dim_ff: dimension of feedforward layer, PyTorch name: dim_feedforward
+    :param dropout: Dropout value, PyTorch name: dropout
+    :param att_dropout: dropout value for attention
+    :param activation: activation function
+    :param norm_eps: Epsilon value for layer normalization
+    :param norm_first: if ``True`` will perform normalization before other att and ff operations, otherwise after
+    :param norm: normalization function
+    """
+    super().__init__()
+    self.self_attn = nn.SelfAttentionStep(
+      key_dim_total=dim_model, value_dim_total=dim_model, num_heads=num_heads, att_dropout=att_dropout)
+    self.multihead_attn = nn.Attention()
+
+    self.linear_ff = nn.Linear(dim_ff)
+    self.linear_out = nn.Linear(dim_model)
+
+    self.norm = norm
+    self.norm_first = norm_first
+    self.norm_eps = norm_eps
+    self.activation = activation
+    self.dropout = dropout
+    self.att_dropout = att_dropout
+
+  def forward(self, x: nn.LayerRef, *, memory: nn.LayerRef, state: nn.LayerState) -> Tuple[nn.LayerRef, nn.LayerState]:
+    """
+    Two possible forward variants of decoder, defined by self.norm_first, x and memory have shape {B, T, F}
+    """
+    if self.norm_first:
+      x, new_state = x + self._self_attention_block(self.norm(x, epsilon=self.norm_eps), state=state)
```

Comment:
This is wrong. I guess `new_state` comes from `_self_attention_block`? Then you cannot add `x` to the tuple.